ghcide/src/Development/IDE/Core/Shake.hs

```diff
@@ -487,23 +479,23 @@ shakeOpen lspEnv defaultConfig logger debouncer
         positionMapping <- newVar HMap.empty
         knownTargetsVar <- newVar $ hashed HMap.empty
         let restartShakeSession = shakeRestart ideState
-        mostRecentProgressEvent <- newTVarIO KickCompleted
         persistentKeys <- newVar HMap.empty
-        let progressUpdate = atomically . writeTVar mostRecentProgressEvent
         indexPending <- newTVarIO HMap.empty
         indexCompleted <- newTVarIO 0
         indexProgressToken <- newVar Nothing
         let hiedbWriter = HieDbWriter{..}
-        progressAsync <- async $
-            when reportProgress $
-                progressThread optProgressStyle mostRecentProgressEvent inProgress
         exportsMap <- newVar mempty
 
+        progress <- do
+            let (before, after) = if testing then (0,0.1) else (0.1,0.1)
```

Comment:
At a delay of 0, we've definitely reduced the window of the race condition, but it's not gone entirely.

Reply:
Agreed - but it's the best I can think of without introducing a separate code path for tests

Reply:
That seems pretty reasonable. Could we leave a comment in the code here that for testing we'd like to guarantee always sending messages, but instead we do a sleep of 0 and a dose of hope?